File: 2019/Thang03_2019/16_03_2019/kfactor.cpp
#include <iostream>
#include <stdio.h>

using namespace std;
//////

const
    int maxk=1e5+100, maxn=5e6+100;
int k, a, b;
bool prime[maxk];
int d[maxn];
//////

void init_prime()
{
    for (int i=2; i<=k; i++) prime[i]=true;
    for (int i=2; i<=k; i++)
        if (prime[i])
        for (int j=i; j<=k/i; j++) prime[i*j]=false;
}
//////

void process()
{
    for (int i=a; i<=b; i++) d[i-a]=i;

    for (int i=2; i<=k; i++)
        if (prime[i])
        {
            int j=(i-a%i)%i;
            while (j<=b-a)
            {
                while (d[j] % i==0) d[j]/=i;
                j+=i;
            }
        }

    int res=0;
    for (int i=0; i<=b-a; i++)
        if (d[i]==1) res++;
    cout << res;
}
//////

int main()
{
    freopen("kfactor.inp", "r", stdin);
    freopen("kfactor.out", "w", stdout);
    scanf("%d%d%d", &k, &a, &b);
    init_prime();
    process();
}

File: 2019/Thang03_2019/16_03_2019/kfactor.pas
const fi = 'kfactor.inp'; fo = 'kfactor.out';
  MaxK = 100001; MaxD = 5000001;

var prime: array[1..MaxK] of boolean;
  v: array[0..MaxD] of longint;
  k, a, b : longint;
  i, j, d, rest : longint;
  res : longint;
  cou : longint;

BEGIN
  assign(input, fi); reset(input);
  readln(k, a, b);
  close(input);

  fillchar(prime, sizeof(prime),true);
  for i := 2 to trunc(sqrt(k)) do
    if  prime[i] then begin
    j := i*i;
    while j <= k do begin
      prime[j]:= false;
      j := j+i;
    end;
  end;

  for i := 0 to b - a do  v[i] := a + i;

  cou:=0;
  for d := 2 to k do
    if prime[d] then begin
      rest := a mod d;
      if rest > 0 then j := d - rest else j:=0;
      while j <= b - a do begin
        while V[j] mod  d = 0 do  v[j]:= v[j] div d;
        j := j + d;
        inc(cou);
      end;
    end;


  res := 0;
  for i := 0 to b - a do
    if v[i] = 1  then res := res + 1;

  assign(output, fo); rewrite(output);
  writeln(res);
  write(cou);
  close(output);
END.

